### PR TITLE
Addition of subyt to sema

### PR DESCRIPTION
**Pull request by cedricdcc**
handles the following issues:

- #56

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -84,7 +84,7 @@ test-coverage-with-graphdb:  ## runs the standard test-suite for all available i
 check:  ## performs linting on the python code
 	@poetry run black --check --diff .
 	@poetry run isort --check --diff .
-	@poetry run flake8 . --exclude ${FLAKE8_EXCLUDE}
+	@poetry run flake8 . --exclude ${FLAKE8_EXCLUDE} --ignore=E203,W503
```
Comment (cedricdcc) [inline Makefile:87]: W503 line break before binary operator -> doesn't play well with max line length
E203 local black formattor kept undoing changes with format on save -> https://github.com/psf/black/issues/315

File: pyproject.toml [hunk 2]
```diff
@@ -16,6 +16,10 @@ python-dateutil = "^2.9.0.post0"
 setuptools = "*"
 uritemplate = "^4.1.1"
 urllib3 = "^2.2.2"
+requests = "^2.32.3"
+typeguard = "^4.3.0"
+xmlasdict = "^0.2.2"
+pandas = "^2.2.2"
```
Comment (marc-portier) [inline pyproject.toml:22]: pandas is quite a heavy dependency - where does this come from?
we might need it eventually, but I don't see how subyt requires this?

File: tests/subyt/test_conditional.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
+import logging
+import os
+import time
+import unittest
+from pathlib import Path
+
+import pandas as pd
```
Comment (marc-portier) [inline tests/subyt/test_conditional.py:7]: apparently this test (only) is introducing dependency to pandas
--> poetry / project.toml should declare this as a test-only dependency ?

File: pyproject.toml (hunk 2 above)
Comment (marc-portier) [reply pyproject.toml:22]: at first glance only for testing --> so declare the dependency appropriately ?

File: sema/subyt/sources.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        def __repr__(self):
+            return "JsonFileSource('%s')" % os.path.abspath(self._json)
+
+    SourceFactory.register("application/json", JsonFileSource)
+except ImportError:
+    log.warn("Python JSON module not available -- disabling JSON support!")
+
+
+try:
+    import xmlasdict
```
Comment (marc-portier) [inline sema/subyt/sources.py:299]: this ensures the xmlasdict is not a hard requirement --> could be moved to test or dev group !

File: pyproject.toml [hunk 5]
```diff
@@ -16,6 +16,10 @@ python-dateutil = "^2.9.0.post0"
 setuptools = "*"
 uritemplate = "^4.1.1"
 urllib3 = "^2.2.2"
+requests = "^2.32.3"
+typeguard = "^4.3.0"
+xmlasdict = "^0.2.2"
```
Comment (marc-portier) [inline pyproject.toml:21]: xmlasdict dependency is also not required for subyt use - only when you want to deal with xml inputs (which we do, but only inside the tests?)

caveat -- documentation will need to make clear that such usage declares its own dependency on xmlasdict

makes me think (new issue?) we could also register a fake source class that makes that statement explicitely when people try inputtinh files that do require the XMLSource

File: sema/subyt/__main__.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def vars_to_dict(vars: list) -> dict:
+    if not vars:
+        return None
+    # else
+    return {name: value for [name, value] in vars}
+
+
+def enable_logging(args: argparse.Namespace):
```
Comment (marc-portier) [inline sema/subyt/__main__.py:161]: feels like duplicate of load_log_config?

File: sema/subyt/__main__.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+import sys
+
+from sema.commons.log.loader import load_log_config
+
+from .api import Generator, GeneratorSettings, Sink
+from .j2.generator import JinjaBasedGenerator
+from .sinks import SinkFactory
+from .sources import SourceFactory
+
+load_log_config()
```
Comment (marc-portier) [inline sema/subyt/__main__.py:14]: why here and not in `main()` ?

File: sema/subyt/api.py [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+    @abstractmethod
+    def __enter__(self) -> Iterable:
+        """Source context is expected to yield an Iterable"""
+
+    @abstractmethod
+    def __exit__(self, *exc):
+        """Source context cleanup"""
+
+
+# TODO make a pandas source in sources.py
```
Comment (marc-portier) [inline sema/subyt/api.py:56]: remove and replace by having a todo-issue for this

File: sema/subyt/api.py [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+            "description": (
+                "Make all keys lowercase so "
+                "to ignore case in key references"
+            ),
+        },
+        "flatten": {
+            "default": True,
+            "description": (
+                "Flatten hierarchical strcutures"
+                " by making hierarchical key references."
```
Comment (marc-portier) [inline sema/subyt/api.py:74]: rather have the space end the previous string?

File: sema/subyt/api.py [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+        """
+        return ",".join(
+            [
+                "no-" + key if not val else key
+                for (key, val) in self._values.items()
+            ]
+        )
+
+    def __repr__(self) -> str:
+        return "GeneratorSettings('%s')" % self.as_modifier_str()
```
Comment (marc-portier) [inline sema/subyt/api.py:151]: f-string?

best to search for ` % ` throughout? 
good chance to streamline our code-style

File: sema/subyt/api.py [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+
+    def __setattr__(self, key: str, val: bool) -> None:
+        if key in ["_values"]:  # actual props to be handled through super
+            super(GeneratorSettings, self).__setattr__(key, val)
+            return
+        # else  --> dynamic props to be stored in self._values
+        if key not in GeneratorSettings._scheme.keys():
+            raise KeyError(
+                "attribute '%s' not writeable in GeneratorSettings object"
+                % key
```
Comment (marc-portier) [inline sema/subyt/api.py:164]: f-string

File: sema/subyt/api.py [hunk 12]
```diff
@@ -0,0 +1,379 @@
...
+
+    def __setitem__(self, key: str, val: Iterable):
+        assert isinstance(
+            val, Iterable
+        ), "This dict only accepts Iterable objects as value"
+        dict.__setitem__(self, key, val)
+
+    def __repr__(self):
+        dictrepr = dict.__repr__(self)
+        return "%s(%s)" % (type(self).__name__, dictrepr)
```
Comment (marc-portier) [inline sema/subyt/api.py:195]: f-string

File: sema/subyt/__main__.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def main():
+    """
+    The main entry point to this module.
+    """
+    args = get_arg_parser().parse_args()
+
+    enable_logging(args)
+    service = make_service(args)
```
Comment (marc-portier) [inline sema/subyt/__main__.py:182]: precursor of the new style "service" interface introduced by sembench
-- maybe we should consider supporting that way of calling the core process with string-arguments only already?

File: sema/subyt/j2/generator.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    """
+
+    def __init__(self, templates_folder: str = "."):
+        """
+        Builds the generator that produces LD output from datasources
+
+        :param templates_folder: Location of the templates defaults to "."
+        """
+        self._templates_folder = templates_folder
+        ttl_filter = {"ttl": Filters.all()["xsd"]}
```
Comment (marc-portier) [inline sema/subyt/j2/generator.py:22]: do we still want this?
move to sema looks like a good time to get rid off the older ttl syntax?

File: sema/subyt/sinks.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+load_log_config()
+log = logging.getLogger(__name__)
+
+
+def assert_writable(file_path: str, force_output: bool = False):
+    if not force_output:
+        assert not os.path.isfile(file_path), (
+            "File to write '%s' already exists" % file_path
+        )
+    parent_path = os.path.dirname(os.path.abspath(file_path))
```
Comment (marc-portier) [inline sema/subyt/sinks.py:20]: os.path to be replaced with pathlib

File: sema/subyt/sinks.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    if not os.path.exists(parent_path):
+        os.makedirs(parent_path)
+    assert os.access(parent_path, os.W_OK), (
+        "Can not write to folder '%s' for creating new files" % parent_path
```
Comment (marc-portier) [inline sema/subyt/sinks.py:24]: f-string

File: sema/subyt/sinks.py [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    def __init__(self, file_path: str, force_output: bool = False):
+        super().__init__()
+        assert_writable(file_path, force_output)
+        self._file_path = file_path
+        self._force_output = force_output
+        if Path(file_path).exists():
+            self.mtimes = {file_path: os.stat(file_path).st_mtime}
+
+    def __repr__(self):
+        return "SingleFileSink('%s', %s)" % (
```
Comment (marc-portier) [inline sema/subyt/sinks.py:83]: * f-string
* pathlib

File: sema/subyt/sinks.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+    def open(self):
+        pass
+
+    def close(self):
+        pass
+
+    def _add(self, file_path: str, part: str, source_mtime: float = None):
+        sink_mtime = (
+            os.stat(file_path).st_mtime if Path(file_path).exists() else 0
```
Comment (marc-portier) [inline sema/subyt/sinks.py:129]: Path(file_path).stat().st_mtime

File: sema/subyt/sources.py [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+
+import requests
+import validators
+from typeguard import check_type
+
+from sema.commons.log.loader import load_log_config
+
+from .api import Source
+
+load_log_config()
```
Comment (marc-portier) [inline sema/subyt/sources.py:17]: why here?
load_log_config should be triggered by main only?

File: sema/subyt/sinks.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+import os
+from pathlib import Path
+
+from uritemplate import URITemplate, variables
+
+from sema.commons.log.loader import load_log_config
+
+from .api import Sink
+
+load_log_config()
```
Comment (marc-portier) [inline sema/subyt/sinks.py:11]: not here?

File: sema/subyt/api.py [hunk 21]
```diff
@@ -0,0 +1,379 @@
+import itertools
+import logging
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from typing import Callable, Dict
+
+from sema.commons.log.loader import load_log_config
+
+load_log_config()
```
Comment (marc-portier) [inline sema/subyt/api.py:9]: not here

File: tests/discovery/test_url_to_graph.py [hunk 22]
```diff
@@ -55,7 +55,7 @@ def test_download_uri_cases():
         graph = get_graph_for_format(uri, formats=[format])
         assert isinstance(graph, Graph)
         assert len(graph) > 0
-        assert len(graph) == case["length"]
+        assert len(graph) >= case["length"]
```
Comment (marc-portier) [inline tests/discovery/test_url_to_graph.py:58]: motivation?
why should test be made more lenient?  
are we allowing errors/issues to go undetected through this?

File: sema/subyt/subyt.py [hunk 23]
```diff
@@ -0,0 +1,52 @@
+import logging
+
+from sema.commons.log.loader import load_log_config
+
+from .api import GeneratorSettings
+from .j2.generator import JinjaBasedGenerator
+from .sinks import SinkFactory
+from .sources import SourceFactory
+
+load_log_config()
```
Comment (marc-portier) [inline sema/subyt/subyt.py:10]: not here

File: pyproject.toml (hunk 5 above)
Comment (marc-portier) [reply pyproject.toml:21]: note: currently there is a log.warn already about the xml support being not available when the xmlasdict dependency is not met -- as suggested above, we might extend that policy with a more upfront message / fail when XML processing is actually attempted later

File: tests/subyt/templates/01-basic.ttl [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+  Target vocabs:
+    - ex: <https://example.org/01-basic/>
+  Inputs: 
+    - input-file: data.csv
+#}
+
+@prefix ex: <https://example.org/01-basic/{{ME}}> .
+<{{uritexpand("https://vliz.be/code/pysubyt/test/item{#id}",_)}}>
+  a ex:test ;
+  ex:name "{{_.name}}"^^xsd:string;
```
Comment (marc-portier) [inline tests/subyt/templates/01-basic.ttl:17]: consider showing off the modern style `{{_name | xsd('string') }}` in stead

Comment (marc-portier) [reply tests/subyt/templates/01-basic.ttl:17]: and check all other templates for similar opportunities --> grep for `^^xsd:`

File: tests/subyt/templates/02-collection.ttl [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+    - set-file(s): 
+        - data_countries.csv as countries,
+        - data.csv as _
+    - mode: iteration,
+#}
+
+{% if ctrl.isFirst %}
+@prefix ex: <https://example.org/02-collection/> .
+ex:fullset ex:iteration {{ctrl.settings.iteration | ttl("xsd:boolean")}};
+    ex:ctrl-first {{ctrl.isFirst | ttl("xsd:boolean")}};
```
Comment (marc-portier) [inline tests/subyt/templates/02-collection.ttl:19]: replace the historic `| ttl` usage with the new style `| xsd`

again grep around in the samples to get rid of all occurances

File: tests/subyt/test_conditional.py [hunk 26]
```diff
@@ -0,0 +1,103 @@
...
+from conftest import run_single_test
+
+from sema.commons.log.loader import load_log_config
+from sema.subyt import Subyt
+
+load_log_config()
```
Comment (marc-portier) [inline tests/subyt/test_conditional.py:13]: tests should only `load_log_config` through run_single_test call in the `main()`

seems this needs to be applied to all tests

**Review by marc-portier — CHANGES_REQUESTED** [review]
- missing some opportunities to do cleanup of some historic stuff
- also introduces new stuff that should be reconsidered imho (mainly a smart selection of when to load_log_config) 
- introducing unwanted dependencies too
- unsure about the increased leniency on the discovery test

surely would like to see some fix on the dependency management, other things could move to follow up issues if need be

File: pyproject.toml (hunk 5 above)
Comment (cedricdcc) [reply pyproject.toml:21]: Further handling done by #68

File: sema/subyt/__main__.py [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+    # when logconf needs to be read
+
+    with open(args.logconf, "r") as yml_logconf:
+        logging.config.dictConfig(
+            yaml.load(yml_logconf, Loader=yaml.SafeLoader)
+        )
+    log.info(f"Logging enabled according to config in {args.logconf}")
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (cedricdcc) [reply sema/subyt/__main__.py:175]: taken up in #69

File: sema/subyt/api.py (hunk 8 above)
Comment (cedricdcc) [reply sema/subyt/api.py:56]: taken up in #70

File: sema/subyt/j2/generator.py (hunk 14 above)
Comment (cedricdcc) [reply sema/subyt/j2/generator.py:22]: I have no idea what this does so I'm leaving it here for now

File: sema/subyt/__main__.py (hunk 13 above)
Comment (cedricdcc) [reply sema/subyt/__main__.py:182]: taken up in issue #71

File: sema/subyt/j2/generator.py (hunk 14 above)
Comment (marc-portier) [reply sema/subyt/j2/generator.py:22]: this keeps `| ttl` working as it looks up the xsd filter and registers it again as ttl too

**Pull request merged**